From 313d4cedef44883ca4eb49c96a31b270da4b548a Mon Sep 17 00:00:00 2001
From: Zhen Han <zhen.han@intel.com>
Date: Thu, 4 Dec 2014 14:24:36 +0800
Subject: [PATCH] ov2722: Fix a BUG in v1p8 power_ctrl failure

Some CHT-T always fail at booting up due to following BUG.

[   10.739891] BUG: unable to handle kernel NULL pointer dereference at (null)
[   10.748660] IP: [<(null)>]  (null)
...
[   10.916272] Call Trace:
[   10.919011]  [<ffffffffa009d7b1>] ? power_up+0x81/0x1d0 [ov2722]
[   10.925727]  [<ffffffffa009dd8a>] ov2722_probe+0x1da/0x500 [ov2722]
[   10.932734]  [<ffffffffa009dbb0>] ? ov2722_s_power+0x80/0x80 [ov2722]
[   10.939936]  [<ffffffff817e9d81>] i2c_device_probe+0xb1/0x130

This issue is because "dev->platform_data->power_ctrl = NULL",
when handing "v1p8 power_ctrl failure" in ov2722_probe().

Solution:
Using power_ctrl(sd, 0) replaces dev->platform_data->power_ctrl(sd, 0)
to avoid panic. Also, take care the gpio_ctrl(sd, 0).

Change-Id: Ia1a8c7fdcc51b8f6ff88ec59c0c8437f59a19922
Tracked-On: https://jira01.devtools.intel.com/browse/IMINAN-13702
Signed-off-by: Zhen Han <zhen.han@intel.com>
Signed-off-by: Xiaoming Wang <xiaoming.wang@intel.com>
---
 .../camera/drivers/media/i2c/ov2722.c              |    4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/drivers/external_drivers/camera/drivers/media/i2c/ov2722.c b/drivers/external_drivers/camera/drivers/media/i2c/ov2722.c
index 46dab1d..e9c6d58 100644
--- a/drivers/external_drivers/camera/drivers/media/i2c/ov2722.c
+++ b/drivers/external_drivers/camera/drivers/media/i2c/ov2722.c
@@ -800,9 +800,9 @@ static int power_up(struct v4l2_subdev *sd)
 	return 0;
 
 fail_clk:
-	dev->platform_data->gpio_ctrl(sd, 0);
+	gpio_ctrl(sd, 0);
 fail_power:
-	dev->platform_data->power_ctrl(sd, 0);
+	power_ctrl(sd, 0);
 	dev_err(&client->dev, "sensor power-up failed\n");
 
 	return ret;
-- 
1.7.9.5

